### PR TITLE
Implementation of wxFontEnumeration for wxQT

### PR DESCRIPTION
**Pull request by ffa-grahamdawes**
The previous implementation was empty,  this new implementation is capable of enumerating the face names and whether they are fixed width or not.  

Unfortunately, I don't know how to request different font encodings in QT.

### CODE REVIEW

**Comment by oneeyeman1** [comment]
@ffa-grahamdawes ,
From what I can gather Qt uses UTF-8 all around, just like GTK+.
You can change the encoding of the text, but not for the font itself.

**Comment by vadz** [comment]
The C++11 range-for loop needs to be rewritten using C++98 syntax, unfortunately, and it also looks like some header is missing (see Travis CI failure).

**Comment by ffa-grahamdawes** [comment]
I've replaced the range based for loop with code that is now C++ 03 compatible.

File: src/qt/fontenum.cpp [hunk 1]
```diff
@@ -7,13 +7,23 @@
 
 // For compilers that support precompilation, includes "wx.h".
 #include "wx/wxprec.h"
-
 #include "wx/fontenum.h"
+#include <QFontDatabase>
```
Comment (catalinr) [inline src/qt/fontenum.cpp]: This include most probably needs the folder, just like all other Qt includes.

File: src/qt/fontenum.cpp [hunk 2]
```diff
@@ -7,13 +7,23 @@
...
 
-bool wxFontEnumerator::EnumerateFacenames( wxFontEncoding WXUNUSED(encoding),
-                bool WXUNUSED(fixedWidthOnly))
+bool wxFontEnumerator::EnumerateFacenames( wxFontEncoding WXUNUSED(encoding), bool fixedWidthOnly)
 {
-    return false;
+    QFontDatabase fontDatabase;
+    const QStringList allFonts = fontDatabase.families(QFontDatabase::Any);
```
Comment (catalinr) [inline src/qt/fontenum.cpp]:
```suggestion
    const QStringList& allFonts = fontDatabase.families(QFontDatabase::Any);
```

File: src/qt/fontenum.cpp [hunk 3]
```diff
@@ -7,13 +7,23 @@
...
+    for( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i)
```
Comment (catalinr) [inline src/qt/fontenum.cpp]:
```suggestion
    for ( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i )
```

File: src/qt/fontenum.cpp [hunk 4]
```diff
@@ -7,13 +7,23 @@
...
+    {
+        const QString &fontFamily = *i;
```
Comment (catalinr) [inline src/qt/fontenum.cpp]:
```suggestion
        const QString& fontFamily = *i;
```

File: src/qt/fontenum.cpp [hunk 5]
```diff
@@ -7,13 +7,23 @@
 
 // For compilers that support precompilation, includes "wx.h".
 #include "wx/wxprec.h"
-
 #include "wx/fontenum.h"
+#include <QFontDatabase>
 
-bool wxFontEnumerator::EnumerateFacenames( wxFontEncoding WXUNUSED(encoding),
-                bool WXUNUSED(fixedWidthOnly))
+bool wxFontEnumerator::EnumerateFacenames( wxFontEncoding WXUNUSED(encoding), bool fixedWidthOnly)
```
Comment (catalinr) [inline src/qt/fontenum.cpp]: `encoding` will probably need a mapping between `wxFontEncoding` and `QFontDatabase::WritingSystem`.

```suggestion
bool wxFontEnumerator::EnumerateFacenames(wxFontEncoding WXUNUSED(encoding), bool fixedWidthOnly)
```

File: src/qt/fontenum.cpp [hunk 6]
```diff
@@ -7,13 +7,23 @@
...
 {
-    return false;
+    QFontDatabase fontDatabase;
+    const QStringList allFonts = fontDatabase.families(QFontDatabase::Any);
+    for( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i)
+    {
+        const QString &fontFamily = *i;
+        if ( !fixedWidthOnly || fontDatabase.isFixedPitch(fontFamily) )
+        {
+            this->OnFacename(fontFamily.toStdString());
```
Comment (catalinr) [inline src/qt/fontenum.cpp]: - `std::string` is not always implicitly convertible to `wxString`; it should use the conversion helper;
- `OnFacename()` can return false, in which case enumeration should break, see the docs;
- `this->` is not used by default in wxW sources.

**Review by catalinr — CHANGES_REQUESTED** [review]
Handling facenames could be improved, and style related issues.

File: src/qt/fontenum.cpp (hunk 6 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: What's the safe way to convert from std::string to wxString?

File: src/qt/fontenum.cpp (hunk 2 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: QFontDatabase::families returns a value not a reference.  We don't indent to modify this I've made it immutable.

File: src/qt/fontenum.cpp (hunk 3 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: I'll apply these changes when I fix up the include paths and build error

File: src/qt/fontenum.cpp (hunk 4 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: I'll apply these changes when I fix up the include paths and build error

File: src/qt/fontenum.cpp (hunk 5 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: I'll apply these changes when I fix up the include paths and build error

File: src/qt/fontenum.cpp (hunk 6 above)
Comment (catalinr) [reply src/qt/fontenum.cpp]: > What's the safe way to convert from std::string to wxString?

You need to convert from `QString` to `wxString` (search for `wxQtConvertString`). You don't need `std::string` at all.

File: src/qt/fontenum.cpp (hunk 2 above)
Comment (catalinr) [reply src/qt/fontenum.cpp]: > QFontDatabase::families returns a value not a reference.

Right, and you don't need to make a copy of it, just keep a const reference to it.

File: src/qt/fontenum.cpp (hunk 6 above)
Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: Cool,  thank you.  I'll implement that change.

File: src/qt/fontenum.cpp [hunk 7]
```diff
@@ -7,13 +7,29 @@
...
+    QFontDatabase fontDatabase;
+    const QStringList allFonts = fontDatabase.families(QFontDatabase::Any);
+    for( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i )
+    {
+        const QString& fontFamily = *i;
+        if ( !fixedWidthOnly || fontDatabase.isFixedPitch(fontFamily) )
+        {
+            if ( !OnFacename(wxQtConvertString(fontFamily)) )
+            {
+                return false;
```
Comment (catalinr) [inline src/qt/fontenum.cpp]: I wonder if it shouldn't just `break;` here.
Other implementations are not completely consistent either, but `break` is used for most of them:
- `dfb` will break and return `true` if at least one facename was found;
- `msw` doesn't use `OnFacename()` AFAICS and will always return true;
- `osx` and `unix` will break and will always return true.

Comment (ffa-grahamdawes) [reply src/qt/fontenum.cpp]: I can't find where the return value of this function is ever used.  Does a value of false mean that no fonts have been found or that there has been some unspecified problem enumerating the fonts?

Comment (catalinr) [reply src/qt/fontenum.cpp]: > I can't find where the return value of this function is ever used.

If you refer to `EnumerateFacenames()`, then indeed, it's not exactly clear what its return value means. But that is not very relevant here, where we care about `OnFacename()` and how its return value should be treated, and new code should at least be consistent with other implementations.

File: src/qt/fontenum.cpp [hunk 8]
```diff
@@ -7,13 +7,29 @@
...
+#include <QtGui/QFontDatabase>
 
-bool wxFontEnumerator::EnumerateFacenames( wxFontEncoding WXUNUSED(encoding),
-                bool WXUNUSED(fixedWidthOnly))
+bool wxFontEnumerator::EnumerateFacenames(wxFontEncoding WXUNUSED(encoding), bool fixedWidthOnly)
 {
-    return false;
+    QFontDatabase fontDatabase;
+    const QStringList allFonts = fontDatabase.families(QFontDatabase::Any);
+    for( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i )
```
Comment (catalinr) [inline src/qt/fontenum.cpp]:
```suggestion
    for ( QStringList::const_iterator i = allFonts.begin(); i != allFonts.end(); ++i )
```

File: src/qt/fontenum.cpp [hunk 9]
```diff
@@ -7,13 +7,29 @@
...
+    {
+        const QString& fontFamily = *i;
+        if ( !fixedWidthOnly || fontDatabase.isFixedPitch(fontFamily) )
+        {
+            if ( !OnFacename(wxQtConvertString(fontFamily)) )
+            {
+               break;
```
Comment (catalinr) [inline src/qt/fontenum.cpp]:
```suggestion
                break;
```

**Review by catalinr — CHANGES_REQUESTED** [review]

**Pull request closed without merging**